fix: insert error handling in list-all & download plugin scripts

laidbackware opened the pull request:
# Summary

This fix will check if a plugin does not exit 0 when calling list-all or download. In the case of an error it will print all stdout and stderr returned by the plugin and then the list all script will exit 1.

Without this fix, it is not possible to test a plugin using CI and get reliable results, as errors are not handled.

Fixes: This fixes https://github.com/asdf-vm/asdf/issues/880 and https://github.com/asdf-vm/asdf/issues/886

## Other Information

This has been tested on Linux and MacOS to be working as expected.

Thank you for contributing to asdf!

Stratus3D commented:
Thanks for the PR! I think there are a couple things we need to do before merging this:

* Write tests for list-all failures, and assert that the command works as expected
* When plugin list-all fails, print a line before the plugin output with something along the lines of `Plugin <plug name>'s list-all callback script failed with output:`

Thanks for the PR! This is a good improvement.

laidbackware commented:
Refactored to include tests - which included fixed 2 existing tests.
I broke some affected string tests out to separate lines to aid readability and prevent long lines.
Added check to install script to allow for plugin download to cleanly exit.
The returncode is collected outside the subshell with no default set prior, as if the download/install block doesn't run the return code of the if command will be captured. The script will then tidy up if either download or install do not exist clearnly.

Stratus3D commented on lib/commands/command-list-all.bash:
```diff
@@ -8,7 +8,15 @@ list_all_command() {
+  if [[ $return_code -ne 0 ]]; then
+    # Printing all output to allow plugin to handle error formatting
+    printf "Plugin %s's list-all callback script failed with output:\\n" "${plugin_name}"
+    printf "%s\\n" "${versions}"
```
I think these prints should go to STDERR instead of STDOUT. Generally only good data should be printed to STDOUT as it is often consumed by other programs.

laidbackware replied:
I've pushed the requested change and all looks good.

Stratus3D commented on test/list_command.bats:
```diff
@@ -26,7 +28,10 @@ teardown() {
+  [[ "$output" == "$(echo -e "dummy\n  1.0.0")"* ]]
+  [[ "$output" == *"$(echo -e "dummy-broken\n  No versions installed")"* ]]
+  [[ "$output" == *"$(echo -e "mummy\n  No versions installed")"* ]]
+  [[ "$output" == *"$(echo -e "tummy\n  2.0.0")" ]]
```
Why does this test code need to change?

laidbackware replied:
Adding the broken plugin caused this test to fail, as the test was expecting an explicit output, which had changed due as the `asdf list` command now includes the broken plugin. I also chose to split over multiple lines to aid readabiliy.

Stratus3D commented:
@laidbackware changes look good! I did leave two comments, one about printing failure output to STDERR, and the other is a question about the tests. Let me know if you will address these comments or if you would like me to make the STDERR change before merging.

laidbackware commented:
@Stratus3D I have also been thinking about another edge case which I'd be interested on your thoughts on before implementing?

This change is currently written to redirect all STDOUT and STDERR data to a single variable when calling list-all. This is fine if the plugin generates an error, but may cause issues if a plugin uses a tool which happens to write to STDERR under normal operation as some linux tools do.
https://github.com/asdf-vm/asdf/blob/e2c4f7c631e4ea242aad46623c936419001da99c/lib/commands/command-list-all.bash#L12
In this case it would skew the output of the asdf command.

The only clean solution I can think of to split the output is to acutally call the list-all script twice. Once just capturing STDOUT which is printed if no errrors occur, but if there is an error it is run a second time to capture STDOUT and STDERR.

What do you think?

jthegedus reviewed (approved):
LGTM. @Stratus3D ?

laidbackware commented:
@Stratus3D @jthegedus I think I've solved the issue I mentioned previously without calling the plugin twice. I was concerned that if there were network timeouts, calling the plugin twice could be worse for the user. 
Now command-list-all.bash contains a function which splits stdout and stderr into separate variables. When the plugin list-all script exits with code 0, then stderr is ignored and when it exits with a non-zero code stderr then stdout is printed. I've also added tests to ensure stderr is ignored when sucessful and printed when unsuccessful.
All CI tests look good on my end.

jthegedus commented:
Thanks very much for contributing this @laidbackware Really appreciate it :smile:

Stratus3D commented:
Sorry @laidbackware for never responding, looks great :)
